pkg/controller/assignmetadata/assignmetadata_controller.go

```diff
@@ -139,38 +171,107 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+	if err != nil {
+		log.Info("could not get/create pod status object", "error", err)
+		return reconcile.Result{}, err
 	}
+	status.Status.MutatorUID = assignMetadata.GetUID()
+	status.Status.ObservedGeneration = assignMetadata.GetGeneration()
+	status.Status.Errors = nil
 
-	mutator, err := mutation.MutatorForAssignMetadata(assignMetadata)
+	mutator, err := mutators.MutatorForAssignMetadata(assignMetadata)
 	if err != nil {
 		log.Error(err, "Creating mutator for resource failed", "resource", request.NamespacedName)
-		tracker.CancelExpect(assignMetadata)
-		return ctrl.Result{}, err
-	}
-	if !deleted {
-		if err := r.system.Upsert(mutator); err != nil {
-			log.Error(err, "Insert failed", "resource", request.NamespacedName)
-			tracker.TryCancelExpect(assignMetadata)
-		} else {
-			tracker.Observe(assignMetadata)
+		tracker.TryCancelExpect(assignMetadata)
+		status.Status.Errors = append(status.Status.Errors, statusv1beta1.Error{Message: err.Error()})
+		if err2 := r.Update(ctx, status); err != nil {
+			log.Error(err2, "could not update mutator status")
 		}
-	} else {
-		if err := r.system.Remove(mutator.ID()); err != nil {
-			log.Error(err, "Remove failed", "resource", request.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	if err := r.system.Upsert(mutator); err != nil {
+		log.Error(err, "Insert failed", "resource", request.NamespacedName)
+		tracker.TryCancelExpect(assignMetadata)
+		status.Status.Errors = append(status.Status.Errors, statusv1beta1.Error{Message: err.Error()})
+		if err2 := r.Update(ctx, status); err != nil {
+			log.Error(err2, "could not update mutator status")
 		}
-		tracker.CancelExpect(assignMetadata)
+		return reconcile.Result{}, err
 	}
 
+	tracker.Observe(assignMetadata)
+	status.Status.Enforced = true
+
+	if err := r.Update(ctx, status); err != nil {
+		log.Error(err, "could not update mutator status")
+		return reconcile.Result{}, err
+	}
 	return ctrl.Result{}, nil
 }
+
+func (r *Reconciler) getOrCreatePodStatus(mutatorID types.ID) (*statusv1beta1.MutatorPodStatus, error) {
+	statusObj := &statusv1beta1.MutatorPodStatus{}
+	sName, err := statusv1beta1.KeyForMutatorID(util.GetPodName(), mutatorID)
+	if err != nil {
+		return nil, err
+	}
+	key := apiTypes.NamespacedName{Name: sName, Namespace: util.GetNamespace()}
+	if err := r.Get(context.TODO(), key, statusObj); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+	} else {
+		return statusObj, nil
+	}
+	pod, err := r.getPod()
+	if err != nil {
+		return nil, err
+	}
+	statusObj, err = statusv1beta1.NewMutatorStatusForPod(pod, mutatorID, r.scheme)
+	if err != nil {
+		return nil, err
+	}
+	if err := r.Create(context.TODO(), statusObj); err != nil {
+		return nil, err
+	}
+	return statusObj, nil
+}
+
+func (r *Reconciler) defaultGetPod() (*corev1.Pod, error) {
+	// require injection of GetPod in order to control what client we use to
+	// guarantee we don't inadvertently create a watch
+	panic("GetPod must be injected")
```

Comment:
Should we also log here

Reply:
Panic writes a core dump to stderr. Is there something specific you want to see with a log line.

FWIW this is a sanity check that the object is initialized properly, the pod wont run and our unit tests/e2e tests (should) fail if we hit this point

Reply:
yea I was thinking at least log the pod name but yea the pod won't even come up so maybe it's a moot point.